File: packages/dropwell/ios/dropwell/Sources/dropwell/DropwellData.swift
import Foundation

/// Pure data handling for the iOS implementation.
///
/// Nothing here touches UIKit, so the XCTest target reaches all of it without
/// a running app. `UIPasteboard` reports a uniform type identifier rather than
/// a media type, and a consumer that wants one should not have to translate
/// Apple's identifiers in every app.
public enum DropwellData {
  /// Returns the final path component.
  public static func fileNameOf(_ path: String) -> String {
    (path as NSString).lastPathComponent
  }

  /// Media type for a file name, or `nil` when unknown.
  public static func mimeFromFileName(_ fileName: String) -> String? {
    guard let dot = fileName.lastIndex(of: "."),
      dot < fileName.index(before: fileName.endIndex)
    else { return nil }
    switch fileName[fileName.index(after: dot)...].lowercased() {
    case "png": return "image/png"
    case "jpg", "jpeg": return "image/jpeg"
    case "webp": return "image/webp"
    case "gif": return "image/gif"
    case "bmp": return "image/bmp"
    case "pdf": return "application/pdf"
    case "json": return "application/json"
    case "csv": return "text/csv"
    case "txt", "md", "log": return "text/plain"
    default: return nil
    }
  }

  /// Media type for a uniform type identifier, or `nil` when unknown.
  public static func mimeFromTypeIdentifier(_ identifier: String) -> String? {
    switch identifier {
    case "public.png": return "image/png"
    case "public.jpeg": return "image/jpeg"
    case "org.webmproject.webp": return "image/webp"
    case "com.compuserve.gif": return "image/gif"
    case "com.adobe.pdf": return "application/pdf"
    case "public.plain-text", "public.utf8-plain-text": return "text/plain"
    default: return nil
    }
  }

  /// A file name for a pasted payload that arrived without one.
  public static func defaultFileName(forMime mime: String?) -> String {
    switch mime {
    case "image/png": return "pasted-image.png"
    case "image/jpeg": return "pasted-image.jpg"
    case "image/webp": return "pasted-image.webp"
    case "image/gif": return "pasted-image.gif"
    default: return "pasted-file"
    }
  }
}

File: packages/dropwell/android/src/main/kotlin/net/tinyrack/dropwell/DropwellData.kt
package net.tinyrack.dropwell

/**
 * Pure data handling for the Android implementation.
 *
 * Nothing here touches the Android framework, so the JVM unit-test target
 * reaches all of it without an emulator. Android's own answer for a clipboard
 * item's media type is often null — a `file://` URI has no provider to ask —
 * and this is the fallback that keeps a pasted screenshot from arriving as an
 * untyped blob.
 */
internal object DropwellData {
    /** Returns the final path component of a URI or path. */
    fun fileNameOf(value: String): String {
        val withoutQuery = value.substringBefore('?')
        val name = withoutQuery.trimEnd('/').substringAfterLast('/')
        return if (name.isEmpty()) withoutQuery else name
    }

    /** Media type for a file name, or null when unknown. */
    fun mimeFromFileName(fileName: String): String? {
        val dot = fileName.lastIndexOf('.')
        if (dot < 0 || dot == fileName.length - 1) return null
        return when (fileName.substring(dot + 1).lowercase()) {
            "png" -> "image/png"
            "jpg", "jpeg" -> "image/jpeg"
            "webp" -> "image/webp"
            "gif" -> "image/gif"
            "bmp" -> "image/bmp"
            "pdf" -> "application/pdf"
            "json" -> "application/json"
            "csv" -> "text/csv"
            "txt", "md", "log" -> "text/plain"
            else -> null
        }
    }

    /**
     * Picks the media type to report for a clipboard item.
     *
     * A content provider's answer wins when it gives one, because it knows the
     * file and the extension may be missing or wrong. `application/octet-stream`
     * is treated as no answer: providers hand it out when they have not looked.
     */
    fun resolveMime(providerMime: String?, fileName: String): String? {
        if (providerMime != null && providerMime != "application/octet-stream") {
            return providerMime
        }
        return mimeFromFileName(fileName)
    }
}
